text_encoder.go

```diff
@@ -0,0 +1,226 @@
+	enc.addKey(key)
+	switch {
+	case math.IsNaN(val):
+		enc.bytes = append(enc.bytes, "NaN"...)
+	case math.IsInf(val, 1):
+		enc.bytes = append(enc.bytes, "+Inf"...)
+	case math.IsInf(val, -1):
+		enc.bytes = append(enc.bytes, "-Inf"...)
+	default:
+		enc.bytes = strconv.AppendFloat(enc.bytes, val, 'f', -1, 64)
+	}
+}
+
+func (enc *textEncoder) AddMarshaler(key string, obj LogMarshaler) error {
+	enc.addKey(key)
+	enc.bytes = append(enc.bytes, '{')
+	err := obj.MarshalLog(enc)
+	enc.bytes = append(enc.bytes, '}')
+	return err
+}
+
+func (enc *textEncoder) AddObject(key string, obj interface{}) error {
+	enc.AddString(key, fmt.Sprintf("%+v", obj))
+	return nil
+}
+
+func (enc *textEncoder) Clone() Encoder {
+	clone := textPool.Get().(*textEncoder)
+	clone.truncate()
+	clone.bytes = append(clone.bytes, enc.bytes...)
+	clone.timeFmt = enc.timeFmt
+	return clone
+}
+
+func (enc *textEncoder) WriteEntry(sink io.Writer, msg string, lvl Level, t time.Time) error {
+	if sink == nil {
+		return errNilSink
+	}
+
+	final := textPool.Get().(*textEncoder)
+	final.truncate()
+	enc.addLevel(final, lvl)
+	enc.addTime(final, t)
+	enc.addMessage(final, msg)
+
+	if len(enc.bytes) > 0 {
+		final.bytes = append(final.bytes, ' ')
+		final.bytes = append(final.bytes, enc.bytes...)
+	}
+	final.bytes = append(final.bytes, '\n')
+
+	expectedBytes := len(final.bytes)
+	n, err := sink.Write(final.bytes)
+	final.Free()
+	if err != nil {
+		return err
+	}
+	if n != expectedBytes {
+		return fmt.Errorf("incomplete write: only wrote %v of %v bytes", n, expectedBytes)
+	}
+	return nil
+}
+
+func (enc *textEncoder) truncate() {
+	enc.bytes = enc.bytes[:0]
+}
+
+func (enc *textEncoder) addKey(key string) {
+	lastIdx := len(enc.bytes) - 1
+	if lastIdx >= 0 && enc.bytes[lastIdx] != '{' {
+		enc.bytes = append(enc.bytes, ' ')
+	}
+	enc.bytes = append(enc.bytes, key...)
+	enc.bytes = append(enc.bytes, '=')
+}
+
+func (enc *textEncoder) addLevel(final *textEncoder, lvl Level) {
+	final.bytes = append(final.bytes, '[')
+	switch lvl {
+	case DebugLevel:
+		final.bytes = append(final.bytes, 'D')
+	case InfoLevel:
+		final.bytes = append(final.bytes, 'I')
+	case WarnLevel:
+		final.bytes = append(final.bytes, 'W')
+	case ErrorLevel:
+		final.bytes = append(final.bytes, 'E')
+	case PanicLevel:
+		final.bytes = append(final.bytes, 'P')
+	case FatalLevel:
+		final.bytes = append(final.bytes, 'F')
+	default:
+		final.bytes = strconv.AppendInt(final.bytes, int64(lvl), 10)
+	}
+	final.bytes = append(final.bytes, ']')
+}
+
+func (enc *textEncoder) addTime(final *textEncoder, t time.Time) {
+	if enc.timeFmt == "" {
+		return
+	}
+	final.bytes = append(final.bytes, ' ')
+	final.bytes = append(final.bytes, t.Format(enc.timeFmt)...)
+}
+
+func (enc *textEncoder) addMessage(final *textEncoder, msg string) {
+	final.bytes = append(final.bytes, ' ')
+	final.bytes = append(final.bytes, msg...)
+}
+
+// A TextOption is used to set options for a text encoder.
```

Comment:
@prashantv can you open an issue describing the desired changes to the APIs so we make sure to follow up in the future? I'm one of those weird people that actually likes following up on loose ends